src/core/lv2/Lv2Worker.cpp

```diff
@@ -0,0 +1,193 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include <cassert>
+#include <QDebug>
+
+#include "Lv2Worker.h"
+
+#ifdef LMMS_HAVE_LV2
+
+#include "Engine.h"
+
+
+namespace lmms
+{
+
+
+// static wrappers (TODO: avoid those somehow?)
+
+static LV2_Worker_Status staticScheduleWork(LV2_Worker_Schedule_Handle handle,
+	uint32_t size, const void* data)
+{
+	Lv2Worker* worker = static_cast<Lv2Worker*>(handle);
+	return worker->scheduleWork(size, data);
+}
+
+
+
+
+static LV2_Worker_Status
+staticWorkerRespond(LV2_Worker_Respond_Handle handle,
+	uint32_t size, const void* data)
+{
+	Lv2Worker* worker = static_cast<Lv2Worker*>(handle);
+	return worker->respond(size, data);
+}
+
+
+
+
+std::size_t Lv2Worker::bufferSize() const
+{
+	// ardour uses this fixed size for ALSA:
+	return 8192 * 4;
+	// for jack, they use 4 * jack_port_type_get_buffer_size (..., JACK_DEFAULT_MIDI_TYPE)
+	// (possible extension for AudioDevice)
+}
+
+
+
+
+Lv2Worker::Lv2Worker(LV2_Handle handle,
+	const LV2_Worker_Interface* iface,
+	Semaphore& common_work_lock,
+	bool threaded) :
+	m_iface(iface),
+	m_threaded(threaded),
+	m_handle(handle),
+	m_response(bufferSize()),
+	m_requests(bufferSize()),
+	m_responses(bufferSize()),
+	m_requestsReader(m_requests),
+	m_responsesReader(m_responses),
+	m_sem(0),
+	m_workLock(common_work_lock)
+{
+	assert(iface);
+	m_scheduleFeature.handle = static_cast<LV2_Worker_Schedule_Handle>(this);
+	m_scheduleFeature.schedule_work = staticScheduleWork;
+
+	if (threaded) { m_thread = std::thread(&Lv2Worker::workerFunc, this); }
+
+	m_requests.mlock();
+	m_responses.mlock();
+}
+
+
+
+
+Lv2Worker::~Lv2Worker()
+{
+	m_exit = true;
+	if(m_threaded) {
+		m_sem.post();
+		m_thread.join();
+	}
+}
+
+
+
+
+LV2_Worker_Status Lv2Worker::respond(uint32_t size, const void* data)
+{
+	if(!size) { return LV2_WORKER_ERR_UNKNOWN; }
+
+	m_responses.write((const char*)&size, sizeof(size));
+	m_responses.write((const char*)data, size);
+	return LV2_WORKER_SUCCESS;
+}
+
+
+
+
+void Lv2Worker::workerFunc()
+{
+	std::vector<char> buf;
+	uint32_t size;
+	while (true) {
+		m_sem.wait();
+		if (m_exit) break;
+		const std::size_t readSpace = m_requestsReader.read_space();
+		if (readSpace <= sizeof(size)) continue; // (should not happen)
+
+		m_requestsReader.read(sizeof(size)).copy((char*)&size, sizeof(size));
+		Q_ASSERT(readSpace - sizeof(size) >= size);
+		try {
+			buf.resize(size);
+		} catch(...) {
+			qWarning() << "Error: reallocating buffer failed";
+			return;
+		}
```

Comment:
Trying to recover from memory allocation failures is usually a lost cause - it might be better to let the program die if resizing fails. At any rate, this doesn't technically follow the LV2 worker specification:
> The host MUST guarantee that [the data] will be eventually passed to work() if [schedule_work] returns LV2_WORKER_SUCCESS

Reply:
Suggesting to replace this block by

```c++
if(size > buf.size())
	buf.resize(size);
```

This will just throw and cause LMMS to crash if the space is not enough. Also, this will avoid a few allocations if the sizes go up and down all the time. Does this sound right?